receiver/azuremonitorreceiver/internal/metadata/metrics.go

```diff
@@ -0,0 +1,272 @@
+// needs to emit metrics from several resources. Otherwise calling this function is not required,
+// just `Emit` function can be called instead.
+// Resource attributes should be provided as ResourceMetricsOption arguments.
+func (mb *MetricsBuilder) EmitForResource(rmo ...ResourceMetricsOption) {
+	rm := pmetric.NewResourceMetrics()
+	rm.Resource().Attributes().EnsureCapacity(mb.resourceCapacity)
+	ils := rm.ScopeMetrics().AppendEmpty()
+	ils.Scope().SetName("otelcol/azuremonitorreceiver")
+	ils.Scope().SetVersion(mb.buildInfo.Version)
+	ils.Metrics().EnsureCapacity(mb.metricsCapacity)
+	mb.EmitAllMetrics(ils)
+
+	for _, op := range rmo {
+		op(mb.resourceAttributesSettings, rm)
+	}
+	if ils.Metrics().Len() > 0 {
+		mb.updateCapacity(rm)
+		rm.MoveTo(mb.metricsBuffer.ResourceMetrics().AppendEmpty())
+	}
+}
+
+// Emit returns all the metrics accumulated by the metrics builder and updates the internal state to be ready for
+// recording another set of metrics. This function will be responsible for applying all the transformations required to
+// produce metric representation defined in metadata and user settings, e.g. delta or cumulative.
+func (mb *MetricsBuilder) Emit(rmo ...ResourceMetricsOption) pmetric.Metrics {
+	mb.EmitForResource(rmo...)
+	metrics := mb.metricsBuffer
+	mb.metricsBuffer = pmetric.NewMetrics()
+	return metrics
+}
+
+// Reset resets metrics builder to its initial state. It should be used when external metrics source is restarted,
+// and metrics builder should update its startTime and reset it's internal state accordingly.
+func (mb *MetricsBuilder) Reset(options ...metricBuilderOption) {
+	mb.startTime = pcommon.NewTimestampFromTime(time.Now())
+	for _, op := range options {
+		op(mb)
+	}
+}
+
+type metricAzureAbstract struct {
+	data     pmetric.Metric // data buffer for generated metric.
+	capacity int            // max observed number of data points added to the metric.
+}
+
+func (m *metricAzureAbstract) updateCapacity() {
+	if m.data.Gauge().DataPoints().Len() > m.capacity {
+		m.capacity = m.data.Gauge().DataPoints().Len()
+	}
+}
+
+func (m *metricAzureAbstract) init(name, unit string) {
+	m.data.SetName(name)
+	m.data.SetUnit(unit)
+	m.data.SetEmptyGauge()
+	m.data.Gauge().DataPoints().EnsureCapacity(m.capacity)
+}
+
+func (mb *MetricsBuilder) getMetric(resourceMetricID string) (*metricAzureAbstract, bool) {
+	if _, exists := mb.metrics[resourceMetricID]; !exists {
+		return nil, false
+	}
+	return mb.metrics[resourceMetricID], true
+}
+
+func (mb *MetricsBuilder) addMetric(resourceMetricID, logicalMetricID, unit string) (*metricAzureAbstract, error) {
+	if _, exists := mb.metrics[resourceMetricID]; exists {
+		return nil, errors.New("metric already exists")
+	}
+
+	m := &metricAzureAbstract{}
+	m.data = pmetric.NewMetric()
+
+	m.init(logicalMetricID, unit)
+
+	mb.metrics[resourceMetricID] = m
+
+	return mb.metrics[resourceMetricID], nil
+}
+
+func (mb *MetricsBuilder) AddDataPoint(
```

Comment:
Normally we use metadata.yaml to define the metrics the receiver produces and then mdatagen to generate all the metrics builder functions necessary to create those metrics.  I'm late to the conversation, but has this topic been brought up yet?

For a metric scraper component like this I expected to see something like these:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/kubeletstatsreceiver/metadata.yaml
- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/nginxreceiver/metadata.yaml
- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/mysqlreceiver/metadata.yaml

Reply:
Can we just link this? https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/metrics-supported

I don't think we'll have the ability to keep the community version up to date at this scale

Reply:
Ok that makes sense.  Will this component automatically get new metrics or will it need updated?